### PR TITLE
docs/design: Infer the System Timezone of a TiDB cluster via TZ environment variable

### PR DESCRIPTION
**Pull request by zhexuany**
### What problem does this PR solve? 
This PR solves pushing `System` to TiKV sides which causes performance delegation. 

### What is changed and how it works?
doc specifies how this change work. 

### Check List 

Tests 
 - No code

Code changes
- No code
Side effects

Related changes

### CODE REVIEW

**Comment by zhexuany** [comment]
@shenli @coocood

File: docs/adding_tz_env.md [hunk 1]
```diff
@@ -0,0 +1,57 @@
+Proposal: adding TZ as an environment variable 
+
+# Proposal: adding TZ as an environment variable as a necessary condition before starting TiDB
```
Comment (coocood) [inline docs/adding_tz_env.md]: Should remove `as a necessary condition`?

File: docs/adding_tz_env.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
+# Proposal: adding TZ as an environment variable 
+
+- Author(s):  [Zhexuan Yang](www.github.com/zhexuany)
+- Last updated:  2018/09/09
+- Discussion at: Not applicable 
+
+## Abstract
+
+When it comes to time-related calculation, it is hard for the distributed system. This proposal tries to resolve two problems: 1. timezone may be inconsistent across multiple `TiDB` instances, 2. performance degradation casued by pushing `System` down to `TiKV`. The impact of this proposal is changing the way of `TiDB` inferring system's timezone name. Before this proposal, the default timezone name pushed down to TiKV is `System` when session's timezone is not set. After this, TiDB evaluates system's timezone name via `TZ` environment variable and then pushes the real name down to `TiKV` rather than `System`.
```
Comment (shenli) [inline docs/adding_tz_env.md]: If the global timezone is set to UTC and the session's timezone is not set explicitly, we will use UTC.
> the default timezone name pushed down to TiKV is `System` when session's timezone is not set

Comment (shenli) [reply docs/adding_tz_env.md]: s/evaluates/infers/

File: docs/adding_tz_env.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
+# Proposal: adding TZ as an environment variable
```
Comment (shenli) [inline docs/adding_tz_env.md]: Infer the `System Timezone` of a TiDB cluster via TZ environment variable

File: docs/adding_tz_env.md [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+## Background
+
+After we solved the daylight saving time issue, we found the performance degradation of TiKV side. Thanks for the investigation done by engineers from TiKV. The root cause of such performance degradation is that TiKV infers `System` timezone name via a third party lib, which calls a syscall and costs a lot. In our internal benchmark system, after [this PR](https://github.com/pingcap/tidb/pull/6823), our codebase is 1000 times slower than before. We have to address this. 
+
+Another problem needs also to be addressed is the potentially incosistent timezone name across multiple `TiDB` instances. `TiDB` instances may reside at different timezone which could produce incorrect calculation when it comes to time-related calculation. Just getting `TiDB`'s sytem timezone could be be broken. We need find a way to ensure the uniqueness of global timezone name across multiple `TiDB`'s timezone name.
```
Comment (shenli) [inline docs/adding_tz_env.md]: could be be broken -> could be broken

File: docs/adding_tz_env.md [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    * std offset
+    * std offset dst [offset], start[/time], end[/time]
+    * :characters
+
+The std means the IANA timezone name; the offset means timezone offset; the dst indicates the leading timezone having daylight saving time. 
+
+In our case, which means both `TiDB` and `TiKV`, we need care the first and third formats. For answering why we do not need the second format, we need to review how Golang evaluates timezone. In `time` package, the method `LoadLocation` reads tzData from pre-specified sources(directories may contain tzData) and then builds `time. Location` from such tzData which already contains daylight saving time information. 
+
+In this proposal, we suggest setting `TZ` to a valid IANA timezone name which can be read from `TiDB` later. If `TiDB` can't get `TZ` or the supply of `TZ` is invalid, `TiDB` just falls back to evaluates the path of the soft link of `/etc/localtime`. In addition, an warning message telling the user you should set `TZ` properly will be printed. Setting `TZ` can be done in our `tidb-ansible` project, it is also can be done at user side by `export TZ="Asia/Shanghai"`. If both of them are failed, `TiDB` will use `UTC` as timezone name.
```
Comment (shenli) [inline docs/adding_tz_env.md]: to evaluates -> to evaluate

Comment (shenli) [reply docs/adding_tz_env.md]: an warning -> a warning

Comment (shenli) [reply docs/adding_tz_env.md]: If both of them are failed, `TiDB` will use `UTC` as timezone name and print a warning log.

File: docs/adding_tz_env.md [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+Firstly, we need to introduce the `TZ` environment. In POSIX system, the value of `TZ` variable can be one of the following three formats. A detailed description can be found in [this link](http://www.gnu.org/software/libc/manual/html_node/TZ-Variable.html)
+
+    * std offset
+    * std offset dst [offset], start[/time], end[/time]
+    * :characters
+
+The std means the IANA timezone name; the offset means timezone offset; the dst indicates the leading timezone having daylight saving time. 
+
+In our case, which means both `TiDB` and `TiKV`, we need care the first and third formats. For answering why we do not need the second format, we need to review how Golang evaluates timezone. In `time` package, the method `LoadLocation` reads tzData from pre-specified sources(directories may contain tzData) and then builds `time. Location` from such tzData which already contains daylight saving time information.
```
Comment (shenli) [inline docs/adding_tz_env.md]: time. Location -> time.Location

Comment (shenli) [reply docs/adding_tz_env.md]: Could you please add a link for LoadLocation?

File: docs/adding_tz_env.md (hunk 5 above)
Comment (shenli) [inline docs/adding_tz_env.md]: which can be read from `TiDB` later -> which can be read by `TiDB` later

File: docs/adding_tz_env.md [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+In this proposal, we suggest setting `TZ` to a valid IANA timezone name which can be read from `TiDB` later. If `TiDB` can't get `TZ` or the supply of `TZ` is invalid, `TiDB` just falls back to evaluates the path of the soft link of `/etc/localtime`. In addition, an warning message telling the user you should set `TZ` properly will be printed. Setting `TZ` can be done in our `tidb-ansible` project, it is also can be done at user side by `export TZ="Asia/Shanghai"`. If both of them are failed, `TiDB` will use `UTC` as timezone name.
+
+The positive side of this change is resolving performance degradation issue and ensuring the uniqueness of gloabl timezone name in multiple `TiDB` instances. 
+
+The negative side is just adding a config item which is a very small matter and the user probably does not care it if we can take care of it and more importantly guarantee the correctness.
```
Comment (shenli) [inline docs/adding_tz_env.md]: Add a note here: We should mention and emphasise it in the document.

File: docs/adding_tz_env.md [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+It does not have compatibility issue as long as the user deploys by `tidb-ansible`. We may mention this in our release-node and the message printed before tidb quits, which must be easy to understand.
+
+
+## Implementation
+
+The implementation is relatively easy. We just get `TZ` environment from system and check whether it is valid or not. If it is invalid, TiDB evaluates the path of soft link of `/etc/localtime`. 
+In addition, a warning message needs to be printed indicating user has to set `TZ` variable. For example, if `/etc/localtime` links to /usr/share/zoneinfo/Asia/Shanghai, then timezone name should be `Asia/Shangahi`.
+
+In order to ensure the uniqueness of global timezone across multiple `TiDB` instances, we need to write timezone name into `mysql.tidb` under column `system_tz`.
```
Comment (shenli) [inline docs/adding_tz_env.md]: When should we insert/update the value of `system_tz `?

File: docs/adding_tz_env.md [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+## Compatibility
+
+It does not have compatibility issue as long as the user deploys by `tidb-ansible`. We may mention this in our release-node and the message printed before tidb quits, which must be easy to understand.
+
+
+## Implementation
+
+The implementation is relatively easy. We just get `TZ` environment from system and check whether it is valid or not. If it is invalid, TiDB evaluates the path of soft link of `/etc/localtime`. In addition, a warning message needs to be printed indicating user has to set `TZ` variable properly. For example, if `/etc/localtime` links to `/usr/share/zoneinfo/Asia/Shanghai`, then timezone name `TiDB` gets should be `Asia/Shangahi`.
+
+In order to ensure the uniqueness of global timezone across multiple `TiDB` instances, we need to write timezone name into `mysql.tidb` under column `system_tz`.  This cached value can be read once `TiDB` finishs its bootstrap stage. A method `loadLocalStr` can do this job.
```
Comment (shenli) [inline docs/adding_tz_env.md]: How can we complement this variable when upgrading from an old cluster?

Comment (zhexuany) [reply docs/adding_tz_env.md]: Old tidb cluster will read system timezone from TZ when a tidb instance is updating its bootstrap version.

Comment (shenli) [reply docs/adding_tz_env.md]: So we should consider full bootstrap and upgrading process.

Comment (zhexuany) [reply docs/adding_tz_env.md]: I just add some desc of this in compatibility section.

Comment (shenli) [reply docs/adding_tz_env.md]: There is no column named `system_tz`.
> we need to write timezone name into `mysql.tidb` under column `system_tz`.

Comment (zhexuany) [reply docs/adding_tz_env.md]: We need add one.

**Comment by shenli** [comment]
LGTM

**Comment by shenli** [comment]
@coocood PTAL

**Pull request merged**